presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -50,11 +50,11 @@ private ExpressionTreeUtils() {}
         return extractExpressions(nodes, clazz, alwaysTrue());
     }
 
-    private static Predicate<FunctionCall> isAggregationPredicate(Metadata metadata)
+    private static boolean isAggregation(FunctionCall functionCall, Metadata metadata)
     {
-        return ((functionCall) -> (metadata.isAggregationFunction(functionCall.getName())
-                || functionCall.getFilter().isPresent()) && !functionCall.getWindow().isPresent()
-                || functionCall.getOrderBy().isPresent());
+        return ((metadata.isAggregationFunction(functionCall.getName()) || functionCall.getFilter().isPresent())
+                && !functionCall.getWindow().isPresent())
+                || functionCall.getOrderBy().isPresent();
```

Comment:
If I am not mistaken, this is equivalent to:
```
boolean isAggregationFunction = metadata.isAggregationFunction(functionCall.getName());
boolean hasFilter = functionCall.getFilter().isPresent();
boolean hasWindow = functionCall.getWindow().isPresent();
boolean hasOrderBy = functionCall.getOrderBy().isPresent();
return ((isAggregationFunction || hasFilter) && !hasWindow) || hasOrderBy;
```

How come non-aggregate function (according to metadata) is reported as an aggregate when filter is defined? I thought one can define filter _on aggregate functions_.

(No change requested, I verified semantics here are not changed.)

Reply:
This method needs to be revisited at some point. Currently, it extracts aggregation "candidates" based on whether they are aggregations or how they are used would require they *be* aggregations. I guess it has evolved over time with patches over patches, and that's what we ended up with.